# MATLAB and Linear Algebra
As mentioned in the previous section, `SPM` is a software package that is built *on top* of the `MATLAB` scientific computing environment. Although possible to use `SPM` without any further knowledge of `MATLAB`, there is an argument to be made for being at least a *bit* familiar with the `MATLAB` language. Unfortunately, we do not have time on this course for in-depth training on computer programming. However, we can start using `MATLAB` as a way of understanding some of the mathematical theory that we will need going forward. This will do double-duty as a way of getting familiar with `MATLAB`, as well as a way of learning some mathematical concepts in a more interactive fashion.

### The MATLAB Prompt
In order to communicate with `MATLAB`, we use the `MATLAB` prompt. The prompt is signified in the Command Window by the two angled brackets `>>`. These brackets indicate where we can type commands for `MATLAB` to interpret. We type something at the prompt, hit <kbd>ENTER</kbd> and then whatever we typed gets sent to the interpreter. If `MATLAB` understands what we typed, it will return the answer. If `MATLAB` does not understand what we typed, it will return an error.

As an example, if we wanted to use `MATLAB` as a highly over-engineered calculator, we could ask what $3 \times 4$ is by typing

In [2]:
3 * 4

ans = 12

However, if we accidentally typed `$` instead of `*`, we would get

In [4]:
3 $ 4

Invalid text character. Check for unsupported symbol, invisible character, or pasting of non-ASCII characters.

In this instance, MATLAB has return an error because it does not undestand what `$` means in this context. 

## Variables
Before getting into some of the key concepts of linear algebra, we first need to understand some programming fundamentals. One of the most basic characteristics of any programming language is the ability to store data in memory for later use. We do this by using *variables*, which you can think of as boxes with names that we can put data inside to keep for later. As an example, the code below shows the declaration of a variable called `a` and assigning it the number 5.

In [7]:
a = 5;

Notice here that we have used a semi-colon `;` at the end of the line to suppress `MATLAB` printing any output. In declaring the variable `a`, we have created a box that we have called `a` that contains the number `5`. Now, every time we type `a`, `MATLAB` will open this box and use whatever is inside it. As such, if we just type `a` we get 

In [9]:
a

a = 5

but if we do something with the returned value we get

In [11]:
a + 7

ans = 12

`````{admonition} Tip
:class: tip
Whenever a value is not explicitly assigned to a variable, `MATLAB` will automatically assign it to a default variable called `ans` (for *answer*).
`````

Similarly, we could declare another variable with the name `b`

In [13]:
b = 10;

and then add `a` and `b` together

In [15]:
a + b

ans = 15

Although these examples are quite trivial, the power of variables means that we can create code where we do not know the values of `a` and `b`. Perhaps their value depends upon the data we are using? Or perhaps it depends on some other factor that may change each time the code runs? Either way, it does not really matter what value they have if we know that we want to add them together. Thus we can write generic code that can be used irrespective of what the inputs will be.

`````{admonition} Tip
:class: tip
There are rules about naming variables. A variable name cannot start with a number, it cannot contain spaces and the only special character you can use is an underscore. You also need to be aware that `MATLAB` is case-sensitive, meaning that `variable_a` and `VARIABLE_A` will be interpreted as different entities. In addition, remember that a variable can only hold one thing at a time and if you re-assign a variable then the original value will be lost.
`````

## Vectors and Matrices
Now that we have the idea of variables in mind, we can move on to understanding some basic principles of linear algebra. Just to reiterate, these are concepts that we will need going forward on the course. As such, it is important to gain some familiarity at this point, even though the context will be missing. Just try and get a feel for some of these ideas and have a go at feeding the code into `MATLAB`. We will start to see all of this in action as the course progresses.

To begin with, the most fundamental entities in linear algebra are *vectors* and *matrices*. We will be using these a lot on this course, so it is important to both understand what a vector or matrix is, as well as understanding how they can be manipulated.

### Vectors
A vector is simply a list of numbers. For instance

$$
\mathbf{a} = \begin{bmatrix}3 & 4 & 5\end{bmatrix}.
$$

Here, $\mathbf{a}$ is a 3-dimensional row vector with elements 3, 4 and 5. We can refer to the individual elements using *subscripts*. For instance, $a_{1} = 3$, $a_{2} = 4$ and so on. Typically, a vector is written as a **bold** lower-case letter. A vector can also be defined as a *column* of numbers, for instance

$$
\mathbf{b} = 
\begin{bmatrix}
    6 \\  
    7 \\ 
    8 \\
\end{bmatrix}.
$$

Here, $\mathbf{b}$ is a 3-dimensional column-vector with elements $b_{1} = 6$, $b_{2} = 7$ and $b_{3} = 8$. Although the orientation of a vector may seem like a trivial detail, we will soon see that this is very important.

In `MATLAB`, a vector is defined using a list of numbers between square brackets. For instance, we can define the vector $\mathbf{a}$ in `MATLAB` as

In [19]:
a = [3 4 5]

a = 1×3 double
     3     4     5

which tells us that `a` is a $1 \times 3$ vector of [double-precision](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) values. We can access the individual elements using curved brackets

In [21]:
a(1)

ans = 3

or provide a range of indices to return multiple values, using the `:` syntax

In [23]:
a(2:3)

ans = 1×2 double
     4     5

which will return all the values of `a` between index `2` and index `3`. 

We can create a column vector by designating new rows with `;`

In [25]:
b = [6; 7; 8]

b = 3×1 double
     6
     7
     8

and, similarly, we can access individual elements using curved brackets

In [27]:
b(1:2)

ans = 2×1 double
     6
     7

### Matrices
A matrix is simply a collection of vectors. We can either think of a matrix as a collection of column-vectors, or as a collection of row-vectors. Either way, a matrix has multiple rows *and* multiple columns. For instance

$$
\mathbf{A} = 
\begin{bmatrix}
1 & 2 & 3 \\
4 & 5 & 6 \\
\end{bmatrix}
$$

is a $2 \times 3$ matrix consisting of elements $a_{11} = 1$, $a_{12} = 2$ and so on. We can refer to any element generically as $a_{ij}$ for row $i$ and column $j$. Typically, matrices are written as upper-case **bold** letters, to distinguish them from vectors.

In `MATLAB`, we declare matrices using square brackets and a combination of spaces and semi-colons. For example, we could declare $\mathbf{A}$ using

In [29]:
A = [1 2 3; 4 5 6]

A = 2×3 double
     1     2     3
     4     5     6

We can also make this declaration look more like a matrix using the MATLAB line-continuation syntax

In [31]:
A = [1 2 3; ...
     4 5 6];

## Matrix Operations
Beyond being containers for collections of numbers, matrices and vectors are *mathematical objects*, that we can treat in similar fashion to regular numbers. As such, arithmetic operations such as addition or subtraction all have matrix equivalents. However, these equivalents may not be particularly intuitive, as we will see below.

### Addition and Subtraction
Starting with addition and subtraction, two matrices can be added or subtracted if they are the same dimensions. In this case, the operation is defined as the addition or subtraction of each individual element. For instance, for a matrix with $m$ rows and $n$ columns, we can write that

$$
\mathbf{C} = \mathbf{A} + \mathbf{B} =
\begin{bmatrix}
a_{11} + b_{11} & a_{12} + b_{12} & \dots  & a_{1n} + b_{1n} \\
a_{21} + b_{21} & a_{22} + b_{22} & \dots  & a_{2n} + b_{2n} \\
\vdots          & \vdots          & \ddots & \vdots          \\
a_{m1} + b_{m1} & a_{m2} + b_{m2} & \dots  & a_{mn} + b_{mn} \\
\end{bmatrix}
$$

where the ellipses ($\dots$) just mean "continue the pattern along this direction". In this example, the row indices run from $i = 1,\dots,m$ and the column indices run from $j = 1,\dots,n$. We can see this in `MATLAB`

In [33]:
A = [1 2 3; 4 5 6];
B = [2 3 4; 5 6 7];
C = A + B

C = 2×3 double
     3     5     7
     9    11    13

where the result `C` is the same as

In [35]:
C = [1+2 2+3 3+4; 4+5 5+6 6+7]

C = 2×3 double
     3     5     7
     9    11    13

Matrix subtraction works in much the same way, so we have

$$
\mathbf{C} = \mathbf{A} - \mathbf{B}
$$

which, again, is only defined when $\mathbf{A}$ and $\mathbf{B}$ have the same dimensions. For instance

In [37]:
A = [1 4 6; 3 2 3];
B = [2 3 4; 5 6 7];
C = A - B

C = 2×3 double
    -1     1     2
    -2    -4    -4

If we try to add/subtract matrices of different dimensions `MATLAB` will throw an error

In [39]:
A = [1 3 4; 5 6 7; 8 9 6];
B = [1 3; 6 5];
C = A + B

Arrays have incompatible sizes for this operation.

Formally, we say that addition or subtraction is *undefined* for matrices of different dimensions. Thus, we can see that the rules that govern what we can and cannot do with matrices are stricter than we may be used to with regular numbers. 

### Multiplication
Multiplication of matrices is also possible, however, this is a more complex operation than addition/subtraction. To begin with, multiplying a matrix by a single number simply *scales* the matrix by that amount. As such, in the world of linear algebra, single numbers are often called *scalars*. For instance

In [42]:
A = [1 2 3; 4 5 6];
B = 2 * A

B = 2×3 double
     2     4     6
     8    10    12

Unfortunately, that is where the simplicity ends. Let us consider the case of multiplying two matrices. For example

$$
\mathbf{A} = 
\begin{bmatrix}
    1 & 2 & 3 \\
    4 & 5 & 6
\end{bmatrix} \quad
\mathbf{B} = 
\begin{bmatrix}
    7  & 8  \\
    9  & 10 \\
    11 & 12
\end{bmatrix}
$$

You may notice that these are different dimensions, which is something that will be explained soon. To begin with, let us see what `MATLAB` tells us $\mathbf{A} \times \mathbf{B}$ is

In [44]:
A = [1 2 3; 4 5 6];
B = [7 8; 9 10; 11 12];
C = A*B

C = 2×2 double
    58    64
   139   154

You might want to take a moment to ponder this result and see if you can work out how `MATLAB` got there.

To see how this multiplication works, start with the first *row* of $\mathbf{A}$ and the first *column* of $\mathbf{B}$. Each element in the row must be multiplied by each element in the column and then added together. In this example, we do $(1 \times 7) + (2 \times 9) + (3 \times 11) = 58$. This sum then becomes the result for the first element of the new matrix

$$
\begin{bmatrix}
    1           & 2           & 3           \\
    \phantom{4} & \phantom{5} & \phantom{6} \\
\end{bmatrix} \times
\begin{bmatrix}
    7  & \phantom{8}  \\
    9  & \phantom{10} \\
    11 & \phantom{12} \\
\end{bmatrix} =
\begin{bmatrix}
    58            & \phantom{64}  \\
    \phantom{139} & \phantom{154} \\
\end{bmatrix}.
$$

We then do the same thing with the first *row* of $\mathbf{A}$ and the second *column* of $\mathbf{B}$ 

$$
\begin{bmatrix}
    1           & 2           & 3           \\
    \phantom{4} & \phantom{5} & \phantom{6} \\
\end{bmatrix} \times
\begin{bmatrix}
    \phantom{7}  & 8  \\
    \phantom{9}  & 10 \\
    \phantom{11} & 12 \\
\end{bmatrix} =
\begin{bmatrix}
    58            & 64  \\
    \phantom{139} & \phantom{154} \\
\end{bmatrix},
$$

then the second *row* of $\mathbf{A}$ and the first *column* of $\mathbf{B}$ 

$$
\begin{bmatrix}
    \phantom{1} & \phantom{2} & \phantom{3} \\
    4           & 5           & 6 \\
\end{bmatrix} \times
\begin{bmatrix}
    7  & \phantom{8}  \\
    9  & \phantom{10} \\
    11 & \phantom{12} \\
\end{bmatrix} =
\begin{bmatrix}
    58  & 64  \\
    139 & \phantom{154} \\
\end{bmatrix}.
$$

Finally, we do the second *row* of $\mathbf{A}$ and the second *column* of $\mathbf{B}$ 

$$
\begin{bmatrix}
    \phantom{1} & \phantom{2} & \phantom{3} \\
    4           & 5           & 6 \\
\end{bmatrix} \times
\begin{bmatrix}
    \phantom{7}  & 8  \\
    \phantom{9}  & 10 \\
    \phantom{11} & 12 \\
\end{bmatrix} =
\begin{bmatrix}
    58  & 64  \\
    139 & 154 \\
\end{bmatrix}.
$$

If this is the first time you have come across matrix multiplication, this will seem very strange. For a deeper explanation of why this is the case, have a look at the info box below. However, the important thing to know from a practical perspective is that matrix-matrix, matrix-vector, vector-matrix and vector-vector multiplication are only defined when the number of *columns* in $\mathbf{A}$ matches the number of *rows* in $\mathbf{B}$. In addition, the resultant matrix will always have the same number of *rows* as $\mathbf{A}$ and the same number of *columns* as $\mathbf{B}$.

```{admonition} Advanced: Why is Matrix Multiplication So Strange?
:class: dropdown

To understand why matrix multiplication has the definition that it does, we need to first understand that linear algebra is all about *linear equations*. A linear equation is simply an equation where some numbers are multiplied by other numbers and then added together to form a weighted-sum. For example, the equation below is a linear equation

$$
y = a_{1}x_{1} + a_{2}x_{2} + b
$$

where there are two *variables* ($x_{1}$ and $x_{2}$) and three *coefficients* ($a_{1}$, $a_{2}$ and $b$). Each variable is multiplied by its respective coefficient and then everything is added together to form the value of y. The important point here is that the variables can take on any value (hence they can *vary*), whereas the coefficients are usually fixed.

Linear equations are incredibly useful in many branches of science for describing real-world phenomena. Often *multiple* equations are needed to model complex phenomena, which is known as a *system* of linear equations. These equations share the same coefficients, but have different values for the variables. For example, take the following system of 3 equations

$$
y_{1} &= a_{1}x_{11} + a_{2}x_{12} + b \\
y_{2} &= a_{1}x_{21} + a_{2}x_{22} + b \\
y_{3} &= a_{1}x_{31} + a_{2}x_{32} + b
$$

Notice here that each equation has different $x$ values (indicated by the different indices), but the same $a$ and $b$ values. The point of matrices is that they allow this system to be represented compactly *and* provides methods for solving for $a$ and $b$. To see this, we can write the system as

$$
\begin{bmatrix}
y_{1} \\
y_{2} \\
y_{3}
\end{bmatrix}
=
\begin{bmatrix}
x_{11} & x_{12} & 1 \\
x_{21} & x_{22} & 1 \\
x_{31} & x_{32} & 1 \\
\end{bmatrix}
\begin{bmatrix}
a_{1} \\
a_{2} \\
b
\end{bmatrix},
$$

which can be expressed simply as

$$
\mathbf{y} = \mathbf{X}\mathbf{a}.
$$

If we use the definition of matrix multiplication, notice that this recreates the linear equations that define the system. For instance, the first row of $\mathbf{X}$ gives

$$
\begin{bmatrix}
x_{11}           & x_{12}           & 1           \\
\phantom{x_{21}} & \phantom{x_{22}} & \phantom{1} \\
\phantom{x_{31}} & \phantom{x_{32}} & \phantom{1} \\
\end{bmatrix}
\begin{bmatrix}
a_{1} \\
a_{2} \\
b
\end{bmatrix} = 
\begin{bmatrix}
x_{11}a_{1} + x_{12}a_{2} + b \\
\phantom{} \\
\phantom{}
\end{bmatrix}.
$$

You can try this yourself for the remaining two rows. The important point here is that this only works because the rules of matrix multiplication make it work. So, the only way for a system of linear equations to be represented in matrices is by having a more convoluted definition of how multiplication works.

```

### Transposition
As we have seen above, the dimensions of a matrix or vector are very important for determining when matrix multiplication can be performed. In practise, this means that the dimensions of matrices may need to be manipulated. One common approach is an operation called *transposition*, where the rows and the columns of a matrix are swapped. The notation for this is either $\mathbf{M}^{\prime}$ or $\mathbf{M}^{T}$ . So that if

$$
\mathbf{M} = 
\begin{bmatrix}
1 & 2 \\
3 & 4 \\
5 & 6
\end{bmatrix},
$$

the transpose would be

$$
\mathbf{M}^{\prime} = 
\begin{bmatrix}
1 & 3 & 5 \\
2 & 4 & 6
\end{bmatrix}.
$$

You can do this in `MATLAB` using the `'` symbol

In [48]:
M = [1 2; 3 4; 5 6]

M = 3×2 double
     1     2
     3     4
     5     6

In [49]:
M'

ans = 2×3 double
     1     3     5
     2     4     6

### Division
We have now covered matrix addition, subtraction and multiplication. So what about matrix division? Unfortunately, this is where things get complicated again. To start with, let us review a particular way of viewing division of normal numbers. As we should all know, division is simply splitting a number into equal parts. However, we can also view this as a *multiplication*. Specifically, multiplication by the *inverse* of another number. For instance

$$
10 \times \frac{1}{2} = 5
$$

because

$$
10 \times \frac{1}{2} = \frac{10}{1} \times \frac{1}{2} = \frac{10}{2} = 5.
$$

This is known as the *multiplicative inverse* of a number. Formally, the inverse is whatever number can be multiplied by the original number to make 1. For instance, the multiplicative inverse of 2 is $\frac{1}{2}$ because $2 \times \frac{1}{2} = 1$. This is written as

$$
2^{-1} = \frac{1}{2}
$$

This is important because there is no definition of division for matrices. However, we know there is multiplication. As such, we perform division by multiplying one matrix by the *inverse* of another.

So what is the *inverse* of a matrix? Like a regular number, it is a matrix that, when multiplied by the original matrix, equals the matrix equivalent of the number 1. So what is the matrix equivalent of the number 1? This is a special matrix called the *identity* matrix that contains 1's on the diagonal and 0's everywhere else.

$$
\mathbf{I} = 
\begin{bmatrix}
    1 & 0 & \dots & 0 \\
    0 & 1 & \dots & 0 \\
    \vdots & \vdots & \ddots & \vdots \\
    0 & 0 & \dots & 1
\end{bmatrix}
$$

If you multiply any matrix by the identity matrix, it does not change and hence is equivalent to the number 1. You can try this yourself in MATLAB using the `eye(n)` function to create an $n \times n$ identity matrix

In [51]:
I = eye(3)

I = 3×3 double
     1     0     0
     0     1     0
     0     0     1

In [52]:
A = [1 2 3; 4 5 6; 7 8 9];
A * I

ans = 3×3 double
     1     2     3
     4     5     6
     7     8     9

The definition of the inverse of the matrix $\mathbf{A}$ is therefore the matrix that satisfies

$$
\mathbf{A}\mathbf{A}^{-1} = \mathbf{I}.
$$

For example

$$
\mathbf{A} = 
\begin{bmatrix}
    4 & 7 \\
    2 & 6
\end{bmatrix} \quad
\mathbf{A}^{-1} = 
\begin{bmatrix}
    \phantom{-}0.6 & -0.7 \\
    -0.2           & \phantom{-}0.4
\end{bmatrix},
$$

because

$$
\begin{bmatrix}
    4 & 7 \\
    2 & 6
\end{bmatrix} \times
\begin{bmatrix}
    \phantom{-}0.6 & -0.7 \\
    -0.2           & \phantom{-}0.4
\end{bmatrix} = 
\begin{bmatrix}
    1 & 0 \\
    0 & 1
\end{bmatrix}
$$

In terms of finding an inverse, there are a multitude of algorithms available. However, we usually just trust `MATLAB` to do the right thing when we use the `inv()` function. For instance,

In [54]:
A = [4 7; 2 6]

A = 2×2 double
     4     7
     2     6

In [55]:
inv(A)

ans = 2×2 double
    0.6000   -0.7000
   -0.2000    0.4000

In [56]:
A * inv(A)

ans = 2×2 double
    1.0000         0
         0    1.0000

The key thing we need to understand is that there are restrictions on finding an inverse. Firstly, a matrix must be *square* to be invertible. Secondly, even if it is square, not every matrix has an inverse. As such, some matrices are *fundamentally uninvertible*. As an example

In [58]:
A = [1 2 3; 3 3 6; 7 2 9];
inv(A);

The reasons why this happens are beyond the scope of this lesson, but just know that in some cases we will not always be able to find a solution to an equation that contains a matrix inverse. We will see this more later in the course, but for now just known that matrix division is *tricky* and can get us in trouble if we are not careful.

## Element-wise Operations
In the discussion above, you may have intuitively thought that both multiplication and division would be equivalent to addition and subtraction, in the sense that we could take two matrices of the same dimensions and either multiply or divide the individual elements. Although this is not the formal definition of either matrix multiplication or division, we can do this in `MATLAB` using *element-wise* operations. These are usually distinguished in code by using a `.` before the operation symbol.

As an example, if we wanted to multiply the individual elements of two matrices we can use `.*` instead of `*`

In [61]:
A = [1 2 3; 4  5  6];
B = [7 8 9; 10 11 12];
C = A .* B

C = 2×3 double
     7    16    27
    40    55    72

Similarly, to divide the elements we can use `./` instead of `/`

In [63]:
A = [1 2 3; 4  5  6];
B = [7 8 9; 10 11 12];
C = A ./ B

C = 2×3 double
    0.1429    0.2500    0.3333
    0.4000    0.4545    0.5000

We will see how these element-wise operators can be very useful when working with imaging data, where the images themselves are respresented as matrices inside `MATLAB`.